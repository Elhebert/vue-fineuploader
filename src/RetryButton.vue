<template>
  <renderless-retry-button
    :id="id"
    :uploader="uploader"
  >
    <button
      slot-scope="{ retryable, buttonEvents }"
      v-if="!onlyRenderIfRetryable || retryable"
      :disabled="!retryable"
      aria-label="retry"
      v-on="buttonEvents"
    >
      Retry
    </button>
  </renderless-retry-button>
</template>

<script>
import RenderlessRetryButton from './renderless/RetryButton'

export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    onlyRenderIfRetryable: {
      type: Boolean,
      default: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  components: {
    RenderlessRetryButton
  }
}
</script>
