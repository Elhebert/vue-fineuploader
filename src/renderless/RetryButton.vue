<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    retryable: false,
    unmounted: false,
  }),

  mounted() {
    this.uploader.on('complete', this.onComplete)
    this.uploader.on('statusChange', this.onStatusChange)
  },

  beforeDestroy() {
    this.unmounted = true
    this.unregisterEventHandlers()
  },

  methods: {
    onComplete(id, name, response) {
      if (id !== this.id || this.unmounted) {
        return
      }

      if (
        !response.success &&
        !this.isRetryForbidden(response) &&
        !this.retryable
      ) {
        this.retryable = true
        return
      }

      if (response.success && this.retryable) {
        this.retryable = false
        return
      }

      if (this.isRetryForbidden(response) && this.retryable) {
        this.retryable = false
        this.unregisterEventHandlers()
      }
    },

    onStatusChange(id, oldStatus, newStatus) {
      if (
        id !== this.id ||
        this.unmounted ||
        newStatus !== this.status.RETRYING_UPLOAD
      ) {
        return
      }

      this.retryable = false
    },

    onClick() {
      if (this.retryable) {
        this.uploader.methods.retry(this.id)
      }
    },

    unregisterEventHandlers() {
      this.uploader.off('complete', this.onComplete)
      this.uploader.off('statusChange', this.onStatusChange)
    },

    isRetryForbidden(response) {
      const preventRetryResponseProperty =
        (this.uploader.options.retry &&
          this.uploader.options.retry.preventRetryResponseProperty) ||
        'preventRetry'

      return !!response[preventRetryResponseProperty]
    },
  },

  render() {
    return this.$scopedSlots.default({
      buttonEvents: {
        click: event => {
          this.onClick(event)
        },
      },
      retryable: this.retryable,
    })
  }
}
</script>
