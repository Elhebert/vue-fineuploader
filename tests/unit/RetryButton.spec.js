import { mount } from '@vue/test-utils'
import RetryButton from '@/RetryButton.vue'
import FineUploader from 'fine-uploader-wrappers'

describe('RetryButton.vue', () => {
  let wrapper
  let renderless
  let uploader

  beforeEach(() => {
    uploader = new FineUploader({ options: {} })

    wrapper = mount(RetryButton, {
      propsData: { id: 0, uploader },
    })
    renderless = wrapper.vm.$children[0]
  })

  it('does not display retry button by default if upload has not failed', () => {
    renderless.onComplete(0, 'foo.bar', { success: true })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isEmpty()).toBe(true)
  })

  it('disables retry button if upload has not failed', () => {
    wrapper.setProps({ onlyRenderIfRetryable: false })
    renderless.onComplete(0, 'foo.bar', { success: true })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().disabled).toBeDefined()
  })

  it('displays retry button if upload has failed', () => {
    wrapper = mount(RetryButton, {
      propsData: { id: 0, uploader },
    })
    renderless = wrapper.vm.$children[0]

    renderless.onComplete(0, 'foo.bar', { success: false })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().disabled).toBeUndefined()
  })

  it('retries upload if button has been clicked', () => {
    renderless.onComplete(0, 'foo.bar', { success: false })
    uploader.methods.retry = jest.fn()
    wrapper.trigger('click')

    expect(wrapper).toMatchSnapshot()
    expect(uploader.methods.retry).toHaveBeenCalledWith(0)
  })

  it('does not display retry button by default if upload has failed and retries are forbidden (default response property)', () => {
    renderless.onComplete(0, 'foo.bar', { success: false, preventRetry: true })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isEmpty()).toBe(true)
  })

  it('does not display retry button by default if upload has failed and retries are forbidden (custom response property)', () => {
    const uploader = new FineUploader({
      options: { retry: { preventRetryResponseProperty: 'dontDareRetry' } },
    })
    wrapper.setProps({ uploader })
    renderless.onComplete(0, 'foo.bar', { success: false, dontDareRetry: true })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isEmpty()).toBe(true)
  })

  it('disables retry button if upload has failed and retries are forbidden', () => {
    wrapper.setProps({ onlyRenderIfRetryable: false })
    renderless.onComplete(0, 'foo.bar', { success: false, preventRetry: true })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().disabled).toBeDefined()
  })
})
