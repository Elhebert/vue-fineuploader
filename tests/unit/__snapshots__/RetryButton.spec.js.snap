// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RetryButton.vue disables retry button if upload has failed and retries are forbidden 1`] = `
<button disabled="disabled" aria-label="retry">
  Retry
</button>
`;

exports[`RetryButton.vue disables retry button if upload has not failed 1`] = `
<button disabled="disabled" aria-label="retry">
  Retry
</button>
`;

exports[`RetryButton.vue displays retry button if upload has failed 1`] = `
<button aria-label="retry">
  Retry
</button>
`;

exports[`RetryButton.vue does not display retry button by default if upload has failed and retries are forbidden (custom response property) 1`] = ``;

exports[`RetryButton.vue does not display retry button by default if upload has failed and retries are forbidden (default response property) 1`] = ``;

exports[`RetryButton.vue does not display retry button by default if upload has not failed 1`] = ``;

exports[`RetryButton.vue retries upload if button has been clicked 1`] = `
<button aria-label="retry">
  Retry
</button>
`;
